# Data cleaning (Census-Income Data Set)

### Loading the data

In [33]:
import numpy as np
import pandas as pd
import csv

from matplotlib import pyplot as plt

%matplotlib inline

ImportError: No module named backends

Let us load the data as a pandas dataframe.

In [2]:
data_train = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data')
len(data_train)

32560

In [3]:
data_test = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.test')
len(data_test)

16281

### Exploring the data
Now take a look at the dataset and its features:

In [5]:
data_train.head(10)

39          State-gov   77516      Bachelors   13           Never-married  \
0  50   Self-emp-not-inc   83311      Bachelors   13      Married-civ-spouse   
1  38            Private  215646        HS-grad    9                Divorced   
2  53            Private  234721           11th    7      Married-civ-spouse   
3  28            Private  338409      Bachelors   13      Married-civ-spouse   
4  37            Private  284582        Masters   14      Married-civ-spouse   
5  49            Private  160187            9th    5   Married-spouse-absent   
6  52   Self-emp-not-inc  209642        HS-grad    9      Married-civ-spouse   
7  31            Private   45781        Masters   14           Never-married   
8  42            Private  159449      Bachelors   13      Married-civ-spouse   
9  37            Private  280464   Some-college   10      Married-civ-spouse   

         Adm-clerical   Not-in-family   White     Male   2174   0   40  \
0     Exec-managerial         Husband   White     Male      0   0   13   
1   Handlers-cleaners   Not-in-family   White     Male      0   0   40   
2   Handlers-cleaners         Husband   Black     Male      0   0   40   
3      Prof-specialty            Wife   Black   Female      0   0   40   
4     Exec-managerial            Wife   White   Female      0   0   40   
5       Other-service   Not-in-family   Black   Female      0   0   16   
6     Exec-managerial         Husband   White     Male      0   0   45   
7      Prof-specialty   Not-in-family   White   Female  14084   0   50   
8     Exec-managerial         Husband   White     Male   5178   0   40   
9     Exec-managerial         Husband   Black     Male      0   0   80   

    United-States   <=50K  
0   United-States   <=50K  
1   United-States   <=50K  
2   United-States   <=50K  
3            Cuba   <=50K  
4   United-States   <=50K  
5         Jamaica   <=50K  
6   United-States    >50K  
7   United-States    >50K  
8   United-States    >50K  
9   United-States    >50K

In [6]:
data_train.dtypes

39                 int64
 State-gov        object
 77516             int64
 Bachelors        object
 13                int64
 Never-married    object
 Adm-clerical     object
 Not-in-family    object
 White            object
 Male             object
 2174              int64
 0                 int64
 40                int64
 United-States    object
 <=50K            object
dtype: object

### Add column headers to data

Description of fnlwgt (final weight)

The weights on the CPS files are controlled to independent estimates of the civilian noninstitutional population of the US.  

These are prepared monthly for us by Population Division here at the Census Bureau.  

##### We use 3 sets of controls.
##### These are:
          1.  A single cell estimate of the population 16+ for each state.
          2.  Controls for Hispanic Origin by age and sex. 
          3.  Controls by Race, age and sex.

In [7]:
data_train.columns = ['age',
                     'workclass',
                     'final-weight',
                     'education',
                     'education-num',
                     'marital-status',
                     'occupation',
                     'relationship',
                     'race',
                     'sex',
                     'capital-gain',
                     'capital-loss',
                     'hours-per-week',
                     'native-country',
                     'income-level'
                    ]


In [8]:
data_train.dtypes


age                int64
workclass         object
final-weight       int64
education         object
education-num      int64
marital-status    object
occupation        object
relationship      object
race              object
sex               object
capital-gain       int64
capital-loss       int64
hours-per-week     int64
native-country    object
income-level      object
dtype: object

### Investigate the attributes for wrong data, fix the typos in classes’ names, standartidize it’s representation
(so later it would be easier to convert them to boolean cagetorical attributes).

#### Cleaning the data: 
- remove quotes
- trailing whitespaces
- tabs
- lowercase all srings 

In [9]:
import re

def trimAllColumns(df):
    """
    Trim whitespace, tabs and \n from ends of each value across all series in dataframe
    """
    trimStrings_1 = lambda x: re.sub('\s+', ' ', x) if type(x) is str else x
    trimStrings_2 = lambda x: x.strip() if type(x) is str else x
    trimStrings_3 = lambda x: x.lower() if type(x) is str else x
    df = df.applymap(trimStrings_1)
    df = df.applymap(trimStrings_2)
    return df.applymap(trimStrings_3)


# simple example of trimming whitespace from data elements
data_train = trimAllColumns(data_train)
data_train.head()

age         workclass  final-weight  education  education-num  \
0   50  self-emp-not-inc         83311  bachelors             13   
1   38           private        215646    hs-grad              9   
2   53           private        234721       11th              7   
3   28           private        338409  bachelors             13   
4   37           private        284582    masters             14   

       marital-status         occupation   relationship   race     sex  \
0  married-civ-spouse    exec-managerial        husband  white    male   
1            divorced  handlers-cleaners  not-in-family  white    male   
2  married-civ-spouse  handlers-cleaners        husband  black    male   
3  married-civ-spouse     prof-specialty           wife  black  female   
4  married-civ-spouse    exec-managerial           wife  white  female   

   capital-gain  capital-loss  hours-per-week native-country income-level  
0             0             0              13  united-states        <=50k  
1             0             0              40  united-states        <=50k  
2             0             0              40  united-states        <=50k  
3             0             0              40           cuba        <=50k  
4             0             0              40  united-states        <=50k

### Take a look at unique values of each column:

#### age

In [10]:
print(sorted((data_train.age.unique())))

[17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 90]


We conclude that everything is ok with ['age'] column - it's integer and no 0 are present => no missing values

#### workclass

In [11]:
print(sorted((data_train.workclass.unique())))

['?', 'federal-gov', 'local-gov', 'never-worked', 'private', 'self-emp-inc', 'self-emp-not-inc', 'state-gov', 'without-pay']


We conclude that there are missing values in ['workclass'] column

I think that workclass data is important for the final model and I'd like to use that in my predictions, so I'd drop rows where it's = '?'

In [12]:
print(len(data_train))
data_train = data_train[data_train.workclass != '?']
print(len(data_train))

32560
30724


#### final-weight

In [13]:
print(sorted((data_train['final-weight'].unique())))[:10]

[13769, 14878, 18827, 19214, 19302, 19395, 19410, 19491, 19520, 19700]


We conclude that everything is ok with ['final-weight'] column - it's integer and no 0 are present => no missing values

#### education

In [14]:
print(sorted((data_train.education.unique())))

['10th', '11th', '12th', '1st-4th', '5th-6th', '7th-8th', '9th', 'assoc-acdm', 'assoc-voc', 'bachelors', 'doctorate', 'hs-grad', 'masters', 'preschool', 'prof-school', 'some-college']


We conclude that everything is ok with ['education'] column - it's string and no '?' are present => no missing values

#### education-num

In [15]:
print(sorted((data_train['education-num'].unique())))

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16]


We conclude that everything is ok with ['education-num'] column - it's integer and no 0 are present => no missing values

#### marital-status

In [16]:
print(sorted((data_train['marital-status'].unique())))

['divorced', 'married-af-spouse', 'married-civ-spouse', 'married-spouse-absent', 'never-married', 'separated', 'widowed']


We conclude that everything is ok with ['marital-status'] column - it's string and no '?' are present => no missing values

#### occupation

In [17]:
print(sorted((data_train['occupation'].unique())))

['?', 'adm-clerical', 'armed-forces', 'craft-repair', 'exec-managerial', 'farming-fishing', 'handlers-cleaners', 'machine-op-inspct', 'other-service', 'priv-house-serv', 'prof-specialty', 'protective-serv', 'sales', 'tech-support', 'transport-moving']


We conclude that there are missing values in ['occupation'] column

I think that occupation data is important for the final model and I'd like to use that in my predictions, so I'd drop rows where it's = '?'

In [18]:
print(len(data_train))
data_train = data_train[data_train.occupation != '?']
print(len(data_train))

30724
30717


#### relationship

In [19]:
print(sorted((data_train['relationship'].unique())))

['husband', 'not-in-family', 'other-relative', 'own-child', 'unmarried', 'wife']


We conclude that everything is ok with ['relationship'] column - it's string and no '?' are present => no missing values

#### race

In [20]:
print(sorted((data_train['race'].unique())))

['amer-indian-eskimo', 'asian-pac-islander', 'black', 'other', 'white']


We conclude that everything is ok with ['race'] column - it's string and no '?' are present => no missing values

#### sex

In [21]:
print(sorted((data_train['sex'].unique())))

['female', 'male']


We conclude that everything is ok with ['sex'] column - it's string and no '?' are present => no missing values

#### hours-per-week

In [22]:
print(sorted((data_train['hours-per-week'].unique())))

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 70, 72, 73, 74, 75, 76, 77, 78, 80, 81, 82, 84, 85, 86, 87, 88, 89, 90, 91, 92, 94, 95, 96, 97, 98, 99]


We conclude that everything is ok with ['hours-per-week'] column - it's integer and no 0 are present => no missing values

#### native-country

In [23]:
print(sorted((data_train['native-country'].unique())))

['?', 'cambodia', 'canada', 'china', 'columbia', 'cuba', 'dominican-republic', 'ecuador', 'el-salvador', 'england', 'france', 'germany', 'greece', 'guatemala', 'haiti', 'holand-netherlands', 'honduras', 'hong', 'hungary', 'india', 'iran', 'ireland', 'italy', 'jamaica', 'japan', 'laos', 'mexico', 'nicaragua', 'outlying-us(guam-usvi-etc)', 'peru', 'philippines', 'poland', 'portugal', 'puerto-rico', 'scotland', 'south', 'taiwan', 'thailand', 'trinadad&tobago', 'united-states', 'vietnam', 'yugoslavia']


We conclude that there are missing values in ['native-country'] column

Let's see how many values are missing:

In [24]:
print(len(data_train[data_train['native-country'] == '?']))

556


556 - is a big number for the data with 32 560 records, let's keep the rows with missing values and treat it like a seperate class

#### income-level

In [25]:
print(sorted((data_train['income-level'].unique())))

['<=50k', '>50k']


We conclude that everything is ok with ['income-level'] column - it's string and no '?' are present => no missing values

#### Remove duplicates if any

In [26]:
print(len(data_train))
data_train.drop_duplicates()
print(len(data_train))

30717
30717


### There are a lot of missing values in columns capital-gain and capital-loss. 
Investigate what is the best way to deal with them, and apply it.

In [27]:
print('Data length is {}\n'.format(len(data_train)))
print('There are {} missing values in capital-gain'.format(len(data_train[data_train['capital-gain'] == 0])))
print('There are {} missing values in capital-loss'.format(len(data_train[data_train['capital-loss'] == 0])))

Data length is 30717

There are 28129 missing values in capital-gain
There are 29256 missing values in capital-loss


### What to do with missing values? 

I've googled that the good strategy would be: 

<i>"If the missing value is a numeric one, replace it with the mean of the associated attribute."</i>

But in this case <b>more than 90% of data is missing</b>, which means that the <b>current mean values</b> of the capital-gain and capital-loss are <b>not repersentative enough</b> 

##### I'd just drop those columns so it won't influence the rest of the data

In [28]:
data_train = data_train.drop(columns=['capital-gain', 'capital-loss'])
data_train.head()

age         workclass  final-weight  education  education-num  \
0   50  self-emp-not-inc         83311  bachelors             13   
1   38           private        215646    hs-grad              9   
2   53           private        234721       11th              7   
3   28           private        338409  bachelors             13   
4   37           private        284582    masters             14   

       marital-status         occupation   relationship   race     sex  \
0  married-civ-spouse    exec-managerial        husband  white    male   
1            divorced  handlers-cleaners  not-in-family  white    male   
2  married-civ-spouse  handlers-cleaners        husband  black    male   
3  married-civ-spouse     prof-specialty           wife  black  female   
4  married-civ-spouse    exec-managerial           wife  white  female   

   hours-per-week native-country income-level  
0              13  united-states        <=50k  
1              40  united-states        <=50k  
2              40  united-states        <=50k  
3              40           cuba        <=50k  
4              40  united-states        <=50k

## Check continuous attributes for outliersand 
##### if you find any - propose the way to deal with them.



In [32]:
# Define a function for a histogram
def histogram(data, x_label, y_label, title):
    _, ax = plt.subplots()
    ax.hist(data, color = '#539caf')
    ax.set_ylabel(y_label)
    ax.set_xlabel(x_label)
    ax.set_title(title)

# Call the function to create plot
histogram(data = data_train['age']
           , x_label = 'Age'
           , y_label = 'Frequency'
           , title = 'Age distribution')



NameError: global name 'plt' is not defined

# The end.